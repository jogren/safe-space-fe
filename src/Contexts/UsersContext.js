import { createContext } from 'react';

export const UsersContext = createContext([]);

export const Provider = props => {
  
}