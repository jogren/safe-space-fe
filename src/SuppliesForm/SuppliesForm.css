.SuppliesForm_section h2 {
  text-align: center;
}

.itemsList_section {
  display: flex;
  justify-content: space-between;
  padding: 0px 200px;
  margin-bottom: 10px;
  position: relative;
}

.itemsList_quantity-container {
  display: flex;
  flex-direction: column;
}

.quantity-text {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.itemsList_quantity-container button {
  height: 30px;
  width: 30px;
  font-size: 18px;
  font-weight: bold;
}

.itemsList_save-changes {
  height: 50px;
  width: 100px;
  font-size: 12px;
  position: absolute;
  right: 50px;
}

.SuppliesForm-add-item {
  display: flex;
  flex-direction: column;
  margin: auto;
  width: 25%;
}

.consumable-check {
  height: 20px;
}

.SuppliesForm_section article {
  overflow: scroll;
}

.SuppliesForm-add-item label {
  margin-bottom: 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.SuppliesForm-add-item label input {
  height: 25px;
  width: 120px;
  font-size: 14px;
}
.SuppliesForm-add-item button {
  height: 30px;
  width: 200px;
  font-size: 14px;
  margin: auto;
}