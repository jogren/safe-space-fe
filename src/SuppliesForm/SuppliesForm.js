import React, { useContext, useState } from 'react';
import { patchItem, addItem } from '../apiCalls/apiCalls';
import { CurrentCenterContext } from '../Contexts/CurrentCenterContext';
import { ItemsContext } from '../Contexts/ItemsContext';
import PropTypes from 'prop-types';
import './SuppliesForm.css';

const SuppliesForm = () => {
  const { reliefCenter, setreliefCenter } = useContext(CurrentCenterContext);
  const { currentItems, setCurrentItems } = useContext(ItemsContext);
  const [idsToEdit, setIdsToEdit] = useState([]);
  const [quantityInput, setQuantityInput] = useState("");
  const [itemToAdd, setItemToAdd] = useState('');  
  const [errorMsg, setErrorMsg] = useState("");
  const [isConsumable, setConsumable] = useState(false);
  
  const updateQuantity = (id, type) => {
    let updatedItems = currentItems.map(item => {
      if(item.id == id) {
        let updatedQuantity = type === 'decrement' ? item.quantity = item.quantity - 1 : item.quantity = item.quantity + 1
        return {
          id: item.id,
          name: item.name,
          quantity: updatedQuantity,
          __typename: "Item" 
        }
      } else {
        return item
      }
    })
    setCurrentItems(updatedItems);
    let editedItems = [...idsToEdit, id];
    setIdsToEdit([...new Set(editedItems)]);
  }

  const updateQuantityOnInput = (id) => {
    let updatedItems = currentItems.map(item => {
      if (item.id == id) {
        return {
          id: item.id,
          name: item.name,
          quantity: quantityInput,
          __typename: "Item"
        }
      } else {
        return item
      }
    })
    setCurrentItems(updatedItems);
  }
  
  const hasBeenEdited = (id) => {
    return idsToEdit.find(itemId => itemId === id);
  }
  
  const saveChanges = async (item) => {
    let filteredIds = idsToEdit.filter(id => id !== item.id);
    setIdsToEdit(filteredIds)
    if (quantityInput) {
      let itemObj = { ...item, quantity: quantityInput }
      try {
        await patchItem(itemObj, reliefCenter);
        updateQuantityOnInput(item.id)
      } catch (error) {
        setErrorMsg('There was an error saving your changes.')
      }
    } else {
      try {
        await patchItem(item, reliefCenter);
      } catch(error) {
        setErrorMsg('There was an error saving your changes.')
      }
    }
  } 

  const handleQuantityInput = (e, item) => {
    setQuantityInput(e.target.value)
    let editedItems = [...idsToEdit, item.id];
    setIdsToEdit([...new Set(editedItems)]);
    if(!e.target.value) {
      let filteredIds = idsToEdit.filter(id => id !== item.id);
      setIdsToEdit(filteredIds)
    }
  }

  const handleAddItem = (e) => {
    setItemToAdd(e.target.value)
  }

  const submitNewItem = async (e) => {
    e.preventDefault();
    let newItem = {
      name: itemToAdd, 
      quantity: 0, 
      consumable: isConsumable 
    }
    let response  = await addItem(newItem, reliefCenter.id)
    let itemWithId = { ...newItem, id: response}
    setCurrentItems([...currentItems, itemWithId])
    setItemToAdd("")
    setConsumable(false)
  }

  const handleConsumable = () => {
    setConsumable(!isConsumable)
  }
        
  const itemsList = currentItems.map((item, index) => {
    return <section key={index} className="itemsList_section">
      <p>{item.name}</p>
      <div className="itemsList_quantity-container">
        <div className="quantity-text">
          <button disabled={item.quantity === 0} onClick={() => updateQuantity(item.id, 'decrement')}>-</button>
          <p>{item.quantity}</p>
          <button onClick={() => updateQuantity(item.id, 'increment')}>+</button>
        </div>
        <div>
          <input 
            className="quantity_input"
            type="number"
            name="quantityInput"
            placeholder="Enter a value..."
            onChange={(e) => handleQuantityInput(e, item)} />
        </div>
      </div>  
      { hasBeenEdited(item.id) && <button className="itemsList_save-changes" onClick={() => saveChanges(item)}>Save Changes</button> }
    </section>
  })
  
  return (
    <section className="SuppliesForm_section">
      <h2>Inventory for {reliefCenter.name}</h2>
      { errorMsg && <p>There was an error saving your changes, please try again.</p> }
      <article>
        {itemsList}
      </article>
      <form className="SuppliesForm-add-item">
        <label>Add an item:
          <input 
            type="text"
            name="addItem"
            placeholder="Enter item name..."
            value={itemToAdd}
            onChange={handleAddItem}
            autoComplete="off" />
        </label>
        <label>Is the item consumable:
          <img
            className="consumable-check"
            alt="add new item"
            src={
              !isConsumable
                ? require("../assets/unchecked.svg")
                : require("../assets/checked.svg")
            }
            onClick={handleConsumable}
          />
        </label>
        <button disabled={!itemToAdd} onClick={submitNewItem}>Add Item</button>
      </form>
    </section>
  )
}

export default SuppliesForm

SuppliesForm.contextTypes = {
  reliefCenter: PropTypes.object,
  setreliefCenter: PropTypes.func,
  currentItems: PropTypes.arrayOf(PropTypes.object),
  setCurrentItems: PropTypes.func
};
