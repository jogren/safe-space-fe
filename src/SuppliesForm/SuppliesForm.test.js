import React from "react";
import { mount } from "enzyme";
import SuppliesForm from "./SuppliesForm";
import { UsersContext } from '../Contexts/UsersContext';
import { ItemsContext } from '../Contexts/ItemsContext';
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("SuppliesForm", () => {
  let wrapper;
  let UsersContext;
  let ItemsContext;

  beforeEach(() => {
      UsersContext = {
        reliefCenter: {
          name: "Turing Relief Center",
          addressPrint: "1331 17th St Denver CO 80202",
          lat: "39.7508132",
          lng: "-104.9967997",
          phone: "5555555555",
          website: "www.relief-center.com",
          id: "1"
        },
        setreliefCenter: () => {}            
      }
      ItemsContext = {
        currentItems: [
          {id: "1", name: "Diapers", quantity: 10},
          {id: "2", name: "Food", quantity: 10},
          {id: "3", name: "Water", quantity: 11},
          {id: "4", name: "Soap", quantity: 7},
          {id: "5", name: "Toothpaste", quantity: 0}
        ],
        setCurrentItems: () => {}            
      }
      wrapper = mount(
        <SuppliesForm/>,
        { UsersContext },
        { ItemsContext }
      );
    });

  it.skip("should render without crashing", () => {
    expect(wrapper.length).toEqual(1);
  });

  it.skip("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip("should mock context", () => {
    wrapper.setContext({ UsersContext })
    wrapper.setContext({ ItemsContext })
    expect(wrapper).toMatchSnapshot();
  });
})