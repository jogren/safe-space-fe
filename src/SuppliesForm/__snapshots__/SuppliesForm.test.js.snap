// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuppliesForm should match the snapshot 1`] = `undefined`;
